### PR TITLE
Add ToggleLineComment and ToggleMultilineComment service

### PR DESCRIPTION
**Pull request by armanio123**
Implements toggleLineComment and toggleMultilineComment services. Considers jsx as well.

### CODE REVIEW

File: src/server/protocol.ts [hunk 1]
```diff
@@ -904,6 +908,18 @@ namespace ts.server.protocol {
         end: Location;
     }
 
+    export interface TextRange {
```
Comment (amcasey) [inline src/server/protocol.ts]: I'd be pretty surprised to learn we don't already have a type for this.  Also, that type (or types) presumably accounts for the fact that VS Code uses line & column, rather than offsets.

Comment (armanio123) [reply src/server/protocol.ts]: Removed in favor of `FileRangeRequestArgs`.

**Comment by sandersn** [comment]
@armanio123 @amcasey is this supposed to ship in TS 3.9 or 4.0? Is it appropriate to ship it in the RC but not the beta? There's no associated bug so I don't have a way to tell.

**Comment by minestarks** [comment]
This should target 4.0 at this point.

**Comment by sandersn** [comment]
What does this need to go into 4.0 now? 
I can see at least it needs to merge from master.

File: src/server/session.ts [hunk 2]
```diff
@@ -2196,6 +2201,38 @@ namespace ts.server {
             });
         }
 
+        private toggleLineComment(args: protocol.FileRangeRequestArgs): TextChange[] {
+            const { file, project } = this.getFileAndProject(args);
```
Comment (sheetalkamat) [inline src/server/session.ts:2205]: These need to use `getFileAndLanguageServiceForSyntacticOperation`

Comment (armanio123) [reply src/server/session.ts]: Changed.

**Comment by armanio123** [comment]
Found a bug with uncomment that got fixed on the latest commit. This PR is ready to be merged as soon as I get an approval.

File: src/services/services.ts [hunk 3]
```diff
@@ -1977,6 +1977,226 @@ namespace ts {
...
+
+        function toggleLineComment(fileName: string, textRange: TextRange, insertComment?: boolean): TextChange[] {
+            const sourceFile = syntaxTreeCache.getCurrentSourceFile(fileName);
+            const textChanges: TextChange[] = [];
+            const { lineStarts, firstLine, lastLine } = getLinesForRange(sourceFile, textRange);
+
+            let isCommenting = insertComment || false;
+            let leftMostPosition = Number.MAX_VALUE;
+            const lineTextStarts = new Map<number>();
+            const whiteSpaceRegex = new RegExp(/\S/);
```
Comment (uniqueiniquity) [inline src/services/services.ts:1996]: `firstNonWhitespaceCharacterPattern`?

File: src/services/services.ts [hunk 4]
```diff
@@ -1977,6 +1977,226 @@ namespace ts {
...
+                        }
+                    });
+                }
+            }
+
+            return textChanges;
+        }
+
+        function commentSelection(fileName: string, textRange: TextRange): TextChange[] {
+            return toggleLineComment(fileName, textRange, /*insertComment*/ true);
```
Comment (uniqueiniquity) [inline src/services/services.ts:2167]: Today if you trigger comment selection within one line, it adds a multiline comment on that range.
Are we changing that behavior?

File: src/server/session.ts [hunk 5]
```diff
@@ -2641,6 +2678,18 @@ namespace ts.server {
             [CommandNames.ProvideCallHierarchyOutgoingCalls]: (request: protocol.ProvideCallHierarchyOutgoingCallsRequest) => {
                 return this.requiredResponse(this.provideCallHierarchyOutgoingCalls(request.arguments));
             },
+            [CommandNames.ToggleLineCommentFull]: (request: protocol.CommentSelectionRequest) => {
```
[the remark this answers is not in the document]
Comment (armanio123) [reply src/server/session.ts:2681]: I was explicitly requested to remove it. I believe is due to VS being the only one using this service so far. @amcasey can provide more details about this.

[the remark this answers is not in the document]
Comment (amcasey) [reply src/server/session.ts:2681]: I believe VS has started making non-full requests since I made that comment. :smile:

[the remark this answers is not in the document]
Comment (armanio123) [reply src/server/session.ts:2681]: Oh! reverted.

File: src/services/types.ts [hunk 6]
```diff
@@ -486,6 +486,11 @@ namespace ts {
 
         /* @internal */ getNonBoundSourceFile(fileName: string): SourceFile;
 
+        toggleLineComment(fileName: string, textRanges: TextRange): TextChange[];
```
[the remark this answers is not in the document]
Comment (armanio123) [reply src/services/types.ts:489]: It should. Changed.

File: src/services/services.ts [hunk 7]
```diff
@@ -1977,6 +1977,226 @@ namespace ts {
...
+                const lineText = sourceFile.text.substring(lineStarts[i], sourceFile.getLineEndOfPosition(lineStarts[i]));
+
+                // Find the start of text and the left-most character. No-op on empty lines.
+                const regExec = whiteSpaceRegex.exec(lineText);
+                if (regExec) {
+                    leftMostPosition = Math.min(leftMostPosition, regExec.index);
+                    lineTextStarts.set(i.toString(), regExec.index);
+
+                    if (lineText.substr(regExec.index, openComment.length) !== openComment) {
+                        isCommenting = insertComment !== undefined ? insertComment : true;
```
[the remark this answers is not in the document]
Comment (armanio123) [reply src/services/services.ts:2011]: Looks cleaner. So changed.

File: src/services/services.ts (hunk 4 above)
Comment (armanio123) [reply src/services/services.ts:2167]: Currently if you run `commentSelection` (Ctrl + K, Ctrl + C), everything is comment as single line. Except if you have a selection in the same line, in this case is a multine comment.

I have fixed this already. Check the test file `commentSelection1.ts`, to see the result.

File: src/services/services.ts (hunk 3 above)
Comment (armanio123) [reply src/services/services.ts:1996]: I'll negotiate to `firstNonWhitespaceCharacterRegex` =).

**Comment by armanio123** [comment]
Thanks @uniqueiniquity for the review. I have addressed all of the changes. Let me know if any additional comments.

File: tests/cases/fourslash/uncommentSelection1.ts [hunk 8]
```diff
@@ -0,0 +1,38 @@
+// Simple comment selection cases.
+
+//// //let var1[| = 1;
```
Comment (amcasey) [inline tests/cases/fourslash/uncommentSelection1.ts:3]: If I uncomment `//// stuff`, I get `// stuff`, right?

File: src/services/utilities.ts [hunk 9]
```diff
@@ -877,14 +877,14 @@ namespace ts {
...
             node.kind === SyntaxKind.ClassKeyword ? isClassDeclaration(parent) || isClassExpression(node) :
-            node.kind === SyntaxKind.FunctionKeyword ? isFunctionDeclaration(parent) || isFunctionExpression(node) :
-            node.kind === SyntaxKind.InterfaceKeyword ? isInterfaceDeclaration(parent) :
-            node.kind === SyntaxKind.EnumKeyword ? isEnumDeclaration(parent) :
-            node.kind === SyntaxKind.TypeKeyword ? isTypeAliasDeclaration(parent) :
-            node.kind === SyntaxKind.NamespaceKeyword || node.kind === SyntaxKind.ModuleKeyword ? isModuleDeclaration(parent) :
-            node.kind === SyntaxKind.ImportKeyword ? isImportEqualsDeclaration(parent) :
-            node.kind === SyntaxKind.GetKeyword ? isGetAccessorDeclaration(parent) :
-            node.kind === SyntaxKind.SetKeyword && isSetAccessorDeclaration(parent)) {
+                node.kind === SyntaxKind.FunctionKeyword ? isFunctionDeclaration(parent) || isFunctionExpression(node) :
```
[the remark this answers is not in the document]
Comment (armanio123) [reply src/services/utilities.ts:880]: It's unintentional in the sense that I didn't add them explicitly, it was added by the formatting command so I decided it was ok. Let me know if you thinkg otherwise.

File: src/services/services.ts [hunk 10]
```diff
@@ -1994,6 +1994,232 @@ namespace ts {
...
+            let isCommenting = insertComment || false;
+            let leftMostPosition = Number.MAX_VALUE;
+            const lineTextStarts = new Map<string, number>();
+            const firstNonWhitespaceCharacterRegex = new RegExp(/\S/);
+            const isJsx = isInsideJsxElement(sourceFile, lineStarts[firstLine]);
+            const openComment = isJsx ? "{/*" : "//";
+
+            // Check each line before any text changes.
+            for (let i = firstLine; i <= lastLine; i++) {
+                const lineText = sourceFile.text.substring(lineStarts[i], sourceFile.getLineEndOfPosition(lineStarts[i]));
```
[the remark this answers is not in the document]
Comment (armanio123) [reply src/services/services.ts:2019]: Technically yes, but there are some considerations that the method already implements. e.g. the array is out of bounds.

File: src/services/services.ts [hunk 11]
```diff
@@ -1994,6 +1994,232 @@ namespace ts {
...
+                    else if (isCommenting) {
+                        textChanges.push({
+                            newText: openComment,
+                            span: {
+                                length: 0,
+                                start: lineStarts[i] + leftMostPosition
+                            }
+                        });
+                    }
+                    else if (sourceFile.text.substr(lineStarts[i] + lineTextStart, openComment.length) === openComment) {
```
[the remark this answers is not in the document]
Comment (armanio123) [reply src/services/services.ts:2051]: `substr` uses the second argument as length. Confirmed.

File: tests/cases/fourslash/commentSelection1.ts [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+//// //let var6[| = 6;
+//// //let var7 = 7;
+//// //let var8 |]= 8;
+
+verify.commentSelection(
+    `//let var1 = 1;
+//let var2 = 2;
+//let var3 = 3;
+
+let var4/* = 4;*/
```
[the remark this answers is not in the document]
Comment (armanio123) [reply tests/cases/fourslash/commentSelection1.ts:20]: We're following commenting implementation of VS. If we feel strongly about it I wouldn't mind changing it back, it would make the code a little bit cleaner.

[the remark this answers is not in the document]
Comment (amcasey) [reply tests/cases/fourslash/commentSelection1.ts:20]: No, it should keep doing whatever it was doing.

File: tests/cases/fourslash/toggleMultilineComment1.ts [hunk 13]
```diff
@@ -0,0 +1,26 @@
+// Simple multiline comment and uncomment.
+
+//// let var1[| = 1;
+//// let var2 = 2;
+//// let var3 |]= 3;
+////
+//// let var4/* = 1;
+//// let var5 [||]= 2;
```
[the remark this answers is not in the document]
Comment (armanio123) [reply tests/cases/fourslash/toggleMultilineComment1.ts:8]: It would just add multiline comment `/**/`. Btw, testing this I found an bug. Fixed and added tests for more edge cases. Check `toggleMultilineComment9.ts`.

File: tests/cases/fourslash/toggleMultilineComment2.ts [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+//// let var8 = 2;
+//// let var9 */= 3;|]
+////
+//// /*let va[|r10 = 1;*/
+//// let var11 = 2;
+//// /*let var12|] = 3;*/
+
+verify.toggleMultilineComment(
+    `let var1/* = 1;
+let var2 *//*= 2;
```
[the remark this answers is not in the document]
Comment (armanio123) [reply tests/cases/fourslash/toggleMultilineComment2.ts:22]: Yeah, this is one of those scenarios that are difficult to decide the best approach. 

IMO, the only reason to not merge the comments is because when we want to reverse the operation we might get a different result:. For example when using jsdoc:
```js
[|const foo = 1;

/** some documentation 
* more doc.
*/
function bar()  {}|]
```
this will result on:
```js
/*const foo = 1;

*//** some documentation 
* more doc.
*//*
function bar()  {}*/
```
Looks messy but in a way it respects existing comments. If you want to uncomment either `foo` or `bar` it should be ok. If we merge comments, the jsdoc syntax will be all mess up.

This is not only for jsdoc, pretty much any existing comment will experience this issue when uncommenting.

File: tests/cases/fourslash/uncommentSelection1.ts (hunk 8 above)
Comment (armanio123) [reply tests/cases/fourslash/uncommentSelection1.ts:3]: Correct, I have added the test to be sure.

File: src/services/utilities.ts [hunk 15]
```diff
@@ -1320,6 +1320,33 @@ namespace ts {
...
+                    || node.kind === SyntaxKind.LessThanToken
+                    || node.kind === SyntaxKind.GreaterThanToken
+                    || node.kind === SyntaxKind.Identifier
+                    || node.kind === SyntaxKind.CloseBraceToken
+                    || node.kind === SyntaxKind.OpenBraceToken
+                    || node.kind === SyntaxKind.SlashToken) {
+                    node = node.parent;
+                }
+                else if (node.kind === SyntaxKind.JsxElement) {
+                    return position > node.getStart(sourceFile) || isInsideJsxElementRecursion(node.parent);
```
[the remark this answers is not in the document]
Comment (armanio123) [reply src/services/utilities.ts:1337]: Good catch, we can. I have change it as well as the name because is no longer make sense.
